### ISSUE TITLE
Regression in v3.5.0: ko.contextFor does not return a context for newly added nodes

### ISSUE THREAD

**Issue by San4es** [knockout/knockout#2445]
We have prepared simple samples to demonstrate the issue:
v3.5.0 - https://jsfiddle.net/2g6k5una/
v3.4.2 - https://jsfiddle.net/2g6k5una/1/

**Comment by fastfasterfastest** [comment]
Isn't the problem that ko.contextFor does _not_ return a context?

**Comment by chrisknoll** [comment]
I'm looking at this, I thought it was understood that unless applyBindings makes the element, there will not be a contextFor it.  In the filddle, apply binding is called but then the div is appended...so if you contextFor on the div, you will get nothing.

**Comment by fastfasterfastest** [comment]
I was merely trying to clarify the issue the OP is reporting - OP reported that `ko.contextFor` returns a context with no `$root` property; I think the issue the OP has is that `ko.contextFor` does not return a context (and yes, this was previously reported in https://github.com/knockout/knockout/issues/2297 and was closed.)

The [documentation](https://knockoutjs.com/documentation/unobtrusive-event-handling.html) for `ko.contextFor` states: "ko.contextFor(element) - returns the entire binding context that was available to the DOM element."  One could argue that in the sample the OP provides, where a DOM element is dynamically added after bindings has been applied, that there was/is no binding context available to the DOM element.

**Comment by chrisknoll** [comment]
Thanks @fastfasterfastest , i wasn't challenging anything you said, just offering a possible backwards-compatible behavior that would make contextFor continue to work, but, as you point out, the 3.4 behavior isn't consistent with the documentation, and I agree getting the code to align with the specification in the docs should be paramount.

Consider a case, however, where you have a custom databinding that generates SVG elements (and this databinding states `controllsDesendantBindings:true`.  If i set up an `event` data-binding to listen for a selector on a svg element (the user clicks on the svg element, raising a click event with the target of the svg element), I think with the 3.5 behavior, i won't be able to get the context for the dynamic SVG element that was clicked using ko.contextFor(e.target)?  

But I could be mis-understanding something: in the fiddle the element was added via `document.appendChild()` and a library like d3 might do the same thing (if I was using a custom binding that added elements to an SVG element via d3).  Do you think child elements created by a custom binding would not result in a non-undefined object when calling context?  If so, that's kinda sad because I do create SVG elements in custom bindigns via d3 and i have it respond to hover events that I then want to get the context for the containing view model from it.

**Comment by fastfasterfastest** [comment]
Obviously the behavior is different in 3.5.0 vs. what it was in 3.4.2.  And to be strictly backward compatible the behavior should not change. But it can be argued that 3.4.2's behavior was incorrect and thus a bug was fixed.

Should `ko.contextFor` traverse the node hierarchy upwards to find the first node that has a binding context, or should you/app do it?  In 3.4.2, `ko.contextFor` did it (and thus returned a context for a DOM element that was arbitrarily added, such as in the OP's sample).  In 3.5.0, you/app must do it.

I think in the use cases you mention, it wouldn't be too hard or tricky in 3.5.0 to get the context, right?  It sits on the element that has the `event` data-binding you mention, where you installed the event handler (perhaps using `ko.contextFor(e.currentTarget)`).

**Comment by chrisknoll** [comment]
Yes, you are right, of course, that each application can implement the same routine to find the closest context.  I'll definitely make a note to have that implemented when we migrate up.  Good idea. thanks.

**Comment by mbest** [comment]
In 3.5.0, `ko.contextFor` was changed to more accurately report the context applied to the DOM node. It will return `undefined` for a DOM node that hasn't been bound by Knockout, even if that node is a child of a node that has been bound. If your goal is to find the context of the nearest parent, you can loop through the parents until you find the context:

```js
function contextForNodeOrNearestParent(node) {
    var context = ko.contextFor(node);
    if (!context && node.parentNode) {
        return contextForNodeOrNearestParent(node.parentNode);
    }
    return context;
}
```

**Comment by San4es** [comment]
Thank you guys for your assistance! We applied the workaround suggested by @mbest!
 
One thing that bothers me - wouldn't it be better to save the previous behavior in the 3.5.0 [minor update](https://semver.org/)? For instance, maybe, you can add a feature toggle or a new API for backward compatibility? This case has worked since v2.0.0: https://jsfiddle.net/rtgoanbL/

**Comment by chrisknoll** [comment]
I've made a PR to add a `findContextFor` function, that I hope can be included in 3.5.1: #2447.

I'm putting this under `ko.utils.findContextFor` unless there's a better place to put it.

**Comment by wahur666** [comment]
Is this shit framework dead yet?
